File: apps/dsa-web/src/types/expectations.ts
// 预期管理系统类型定义

export type IndexDirection = 'up' | 'flat' | 'down'
export type IndexMagnitude = 'strong' | 'moderate' | 'weak'
export type ExpectationMarket = 'cn' | 'hk' | 'us'
export type StockAction = 'buy' | 'sell' | 'add' | 'reduce' | 'hold' | 'watch'
export type StockDirection = 'up' | 'flat' | 'down'
export type ResearchTime = 'lt_30m' | '30_90m' | 'gt_90m'
export type ExecutionStatus = 'executed' | 'partial' | 'not_executed'

export interface StockExpectationItem {
  code: string
  name?: string | null
  action: StockAction
  direction: StockDirection
  targetPrice?: number | null
  stopLoss?: number | null
  confidence: number
  reasoning?: string | null
}

export interface UserExpectation {
  id: number
  targetDate: string
  market: ExpectationMarket
  createdAt?: string | null
  updatedAt?: string | null

  // 心理快照
  emotionIndex?: number | null
  decisionDrivers?: string[] | null
  researchTime?: ResearchTime | null
  interferenceFlags?: string[] | null

  // 大盘预期
  indexDirection: IndexDirection
  indexMagnitude?: IndexMagnitude | null
  indexReasoning: string

  // 个股 & 计划
  stockExpectations?: StockExpectationItem[] | null
  keyAssumptions?: string[] | null
  keyRisks?: string | null
  operationPlan?: string | null
  overallConfidence?: number | null
  tags?: string[] | null
}

export interface ExpectationListResponse {
  items: UserExpectation[]
  total: number
  page: number
  pageSize: number
}

// ---------- Outcome ----------

export interface IndexScoreDetail {
  predictedDirection: string
  predictedMagnitude?: string | null
  actualPctChg?: number | null
  actualDirection?: string | null
  actualMagnitude?: string | null
  directionHit: boolean
  magnitudeHit: boolean
  score: number
  note?: string
}

export interface StockScoreItem {
  code: string
  predictedDirection: string
  actualPctChg?: number | null
  actualDirection?: string | null
  directionHit: boolean
  score: number
  note?: string
}

export interface AssumptionReview {
  assumption: string
  result: 'hit' | 'miss' | 'na'
}

export interface ExpectationOutcome {
  id: number
  expectationId: number
  outcomeDate?: string | null
  scoredAt?: string | null
  autoScore?: number | null
  indexScoreDetail?: IndexScoreDetail | null
  stockScores?: StockScoreItem[] | null
  selfScore?: number | null
  executionStatus?: ExecutionStatus | null
  executionNotes?: string | null
  deviationReason?: string | null
  assumptionReviews?: AssumptionReview[] | null
  lessons?: string | null
  filledAt?: string | null
}

// ---------- Agent 评价 ----------

export interface AgentEval {
  id: number
  expectationId: number
  evalType: 'single' | 'weekly'
  evalDate?: string | null
  generatedAt?: string | null
  reasoningQuality?: number | null
  informationUsage?: number | null
  riskAwareness?: number | null
  executionAlignment?: number | null
  overallAssessment?: string | null
  strengths?: string[] | null
  weaknesses?: string[] | null
  improvementSuggestions?: string[] | null
  biasTags?: string[] | null
}

// ---------- 聚合详情 ----------

export interface ExpectationDetail {
  expectation: UserExpectation
  outcome?: ExpectationOutcome | null
  agentEval?: AgentEval | null
}

// ---------- 请求 ----------

export interface ExpectationCreateRequest {
  targetDate: string
  market?: ExpectationMarket
  emotionIndex?: number | null
  decisionDrivers?: string[] | null
  researchTime?: ResearchTime | null
  interferenceFlags?: string[] | null
  indexDirection: IndexDirection
  indexMagnitude?: IndexMagnitude | null
  indexReasoning: string
  stockExpectations?: StockExpectationItem[] | null
  keyAssumptions?: string[] | null
  keyRisks?: string | null
  operationPlan?: string | null
  overallConfidence?: number
  tags?: string[] | null
}

export interface ExpectationUpdateRequest {
  emotionIndex?: number | null
  decisionDrivers?: string[] | null
  researchTime?: ResearchTime | null
  interferenceFlags?: string[] | null
  indexDirection?: IndexDirection
  indexMagnitude?: IndexMagnitude | null
  indexReasoning?: string
  stockExpectations?: StockExpectationItem[] | null
  keyAssumptions?: string[] | null
  keyRisks?: string | null
  operationPlan?: string | null
  overallConfidence?: number | null
  tags?: string[] | null
}

export interface SelfReviewRequest {
  selfScore?: number | null
  executionStatus?: ExecutionStatus | null
  executionNotes?: string | null
  deviationReason?: string | null
  assumptionReviews?: AssumptionReview[] | null
  lessons?: string | null
}

export interface ExpectationListParams {
  market?: ExpectationMarket
  fromDate?: string
  toDate?: string
  page?: number
  pageSize?: number
}

File: api/v1/schemas/expectations.py
# -*- coding: utf-8 -*-
"""Expectation management API schemas."""

from __future__ import annotations

from datetime import date, datetime
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

IndexDirection = Literal['up', 'flat', 'down']
IndexMagnitude = Literal['strong', 'moderate', 'weak']
ExpectationMarket = Literal['cn', 'hk', 'us']
StockAction = Literal['buy', 'sell', 'add', 'reduce', 'hold', 'watch']
StockDirection = Literal['up', 'flat', 'down']
ResearchTime = Literal['lt_30m', '30_90m', 'gt_90m']
ExecutionStatus = Literal['executed', 'partial', 'not_executed']


class StockExpectationItem(BaseModel):
    code: str = Field(..., min_length=1, max_length=20)
    name: Optional[str] = None
    action: StockAction
    direction: StockDirection
    target_price: Optional[float] = None
    stop_loss: Optional[float] = None
    confidence: int = Field(3, ge=1, le=5)
    reasoning: Optional[str] = None


# ---------- 预期主记录 ----------

class ExpectationCreateRequest(BaseModel):
    target_date: date
    market: ExpectationMarket = 'cn'

    # 心理快照（可选）
    emotion_index: Optional[int] = Field(None, ge=1, le=10)
    decision_drivers: Optional[List[str]] = None
    research_time: Optional[ResearchTime] = None
    interference_flags: Optional[List[str]] = None

    # 大盘预期（必填）
    index_direction: IndexDirection
    index_magnitude: Optional[IndexMagnitude] = None
    index_reasoning: str = Field(..., min_length=1)

    # 个股预期
    stock_expectations: Optional[List[StockExpectationItem]] = None

    # 假设与计划
    key_assumptions: Optional[List[str]] = None
    key_risks: Optional[str] = None
    operation_plan: Optional[str] = None
    overall_confidence: int = Field(3, ge=1, le=5)
    tags: Optional[List[str]] = None

    model_config = ConfigDict(extra='forbid')


class ExpectationUpdateRequest(BaseModel):
    emotion_index: Optional[int] = Field(None, ge=1, le=10)
    decision_drivers: Optional[List[str]] = None
    research_time: Optional[ResearchTime] = None
    interference_flags: Optional[List[str]] = None
    index_direction: Optional[IndexDirection] = None
    index_magnitude: Optional[IndexMagnitude] = None
    index_reasoning: Optional[str] = None
    stock_expectations: Optional[List[StockExpectationItem]] = None
    key_assumptions: Optional[List[str]] = None
    key_risks: Optional[str] = None
    operation_plan: Optional[str] = None
    overall_confidence: Optional[int] = Field(None, ge=1, le=5)
    tags: Optional[List[str]] = None

    model_config = ConfigDict(extra='forbid')


class ExpectationItem(BaseModel):
    id: int
    target_date: date
    market: str
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    emotion_index: Optional[int] = None
    decision_drivers: Optional[Any] = None
    research_time: Optional[str] = None
    interference_flags: Optional[Any] = None

    index_direction: str
    index_magnitude: Optional[str] = None
    index_reasoning: str

    stock_expectations: Optional[Any] = None
    key_assumptions: Optional[Any] = None
    key_risks: Optional[str] = None
    operation_plan: Optional[str] = None
    overall_confidence: Optional[int] = None
    tags: Optional[Any] = None

    model_config = ConfigDict(from_attributes=True)


class ExpectationListResponse(BaseModel):
    items: List[ExpectationItem]
    total: int
    page: int
    page_size: int


# ---------- 结果与复盘 ----------

class SelfReviewRequest(BaseModel):
    self_score: Optional[int] = Field(None, ge=1, le=5)
    execution_status: Optional[ExecutionStatus] = None
    execution_notes: Optional[str] = None
    deviation_reason: Optional[str] = None
    assumption_reviews: Optional[List[Dict[str, Any]]] = None
    lessons: Optional[str] = None

    model_config = ConfigDict(extra='forbid')


class OutcomeItem(BaseModel):
    id: int
    expectation_id: int
    outcome_date: Optional[date] = None
    scored_at: Optional[datetime] = None
    auto_score: Optional[float] = None
    index_score_detail: Optional[Any] = None
    stock_scores: Optional[Any] = None
    self_score: Optional[int] = None
    execution_status: Optional[str] = None
    execution_notes: Optional[str] = None
    deviation_reason: Optional[str] = None
    assumption_reviews: Optional[Any] = None
    lessons: Optional[str] = None
    filled_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


# ---------- Agent 评价 ----------

class AgentEvalItem(BaseModel):
    id: int
    expectation_id: int
    eval_type: str
    eval_date: Optional[date] = None
    generated_at: Optional[datetime] = None
    reasoning_quality: Optional[int] = None
    information_usage: Optional[int] = None
    risk_awareness: Optional[int] = None
    execution_alignment: Optional[int] = None
    overall_assessment: Optional[str] = None
    strengths: Optional[Any] = None
    weaknesses: Optional[Any] = None
    improvement_suggestions: Optional[Any] = None
    bias_tags: Optional[Any] = None

    model_config = ConfigDict(from_attributes=True)


# ---------- 聚合详情 ----------

class ExpectationDetailResponse(BaseModel):
    expectation: ExpectationItem
    outcome: Optional[OutcomeItem] = None
    agent_eval: Optional[AgentEvalItem] = None


# ---------- 通用 ----------

class ErrorResponse(BaseModel):
    error: str
    message: str


class ScoreResponse(BaseModel):
    outcome: OutcomeItem
